src/libs/actions/FormActions.ts

```diff
@@ -0,0 +1,26 @@
+import Onyx from 'react-native-onyx';
+import {PartialDeep} from 'type-fest';
+import {KeyValueMapping} from 'react-native-onyx/lib/types';
+import * as OnyxCommon from '../../types/onyx/OnyxCommon';
+import {OnyxKey, OnyxValues} from '../../ONYXKEYS';
+
+type KeysWhichCouldBeDraft<T extends keyof OnyxValues> = T extends `${infer U}Draft` ? U : never;
+type DraftKeysWithoutDraftSuffix = KeysWhichCouldBeDraft<keyof OnyxValues>;
+
+function setIsLoading(formID: OnyxKey, isLoading: boolean) {
+    Onyx.merge(formID, {isLoading});
+}
+
+function setErrors(formID: OnyxKey, errors: OnyxCommon.Errors) {
+    Onyx.merge(formID, {errors});
+}
+
+function setErrorFields(formID: OnyxKey, errorFields: OnyxCommon.ErrorFields) {
+    Onyx.merge(formID, {errorFields});
+}
+
+function setDraftValues<T extends DraftKeysWithoutDraftSuffix>(formID: T, draftValues: PartialDeep<KeyValueMapping[`${T}Draft`]>) {
+    Onyx.merge(`${formID}Draft`, draftValues);
+}
```

Comment:
```suggestion
function setDraftValues<T extends OnyxFormKey | OnyxKeysMap['REIMBURSEMENT_ACCOUNT']>(formID: T, draftValues: PartialDeep<KeyValueMapping[T]>) {
    Onyx.merge(`${formID}Draft` as OnyxKey, draftValues);
}
```

We have an issue here: `setDraftValues()` can be used for all form keys, but the draft version of these keys are not declared in `ONYXKEYS.ts`. We have some options:

1. For each form key, create a "draft" version in `ONYXKEYS.ts`, but it will slightly increase the number of keys there and could make typings more complex and slow.
2. Relax the typings here like I did in my suggestion to accept form keys or `REIMBURSEMENT_ACCOUNT`. Another option is to move `REIMBURSEMENT_ACCOUNT` to inside `ONYXKEYS.FORM` if that is really a form. This way we don't need to create a draft key for every form key in `ONYXKEYS.ts`.


Regardless the option we chose, we need to add some missing Form keys to `OnyxValues`, here is the updated list.

```ts
    // Forms
    [ONYXKEYS.FORMS.ADD_DEBIT_CARD_FORM]: OnyxTypes.AddDebitCardForm;
    [ONYXKEYS.FORMS.WORKSPACE_SETTINGS_FORM]: OnyxTypes.Form;
    [ONYXKEYS.FORMS.WORKSPACE_RATE_AND_UNIT_FORM]: OnyxTypes.Form;
    [ONYXKEYS.FORMS.CLOSE_ACCOUNT_FORM]: OnyxTypes.Form;
    [ONYXKEYS.FORMS.PROFILE_SETTINGS_FORM]: OnyxTypes.Form;
    [ONYXKEYS.FORMS.DISPLAY_NAME_FORM]: OnyxTypes.Form;
    [ONYXKEYS.FORMS.ROOM_NAME_FORM]: OnyxTypes.Form;
    [ONYXKEYS.FORMS.WELCOME_MESSAGE_FORM]: OnyxTypes.Form;
    [ONYXKEYS.FORMS.LEGAL_NAME_FORM]: OnyxTypes.Form;
    [ONYXKEYS.FORMS.WORKSPACE_INVITE_MESSAGE_FORM]: OnyxTypes.Form;
    [ONYXKEYS.FORMS.DATE_OF_BIRTH_FORM]: OnyxTypes.Form;
    [ONYXKEYS.FORMS.HOME_ADDRESS_FORM]: OnyxTypes.Form;
    [ONYXKEYS.FORMS.NEW_ROOM_FORM]: OnyxTypes.Form;
    [ONYXKEYS.FORMS.ROOM_SETTINGS_FORM]: OnyxTypes.Form;
    [ONYXKEYS.FORMS.NEW_TASK_FORM]: OnyxTypes.Form;
    [ONYXKEYS.FORMS.EDIT_TASK_FORM]: OnyxTypes.Form;
    [ONYXKEYS.FORMS.MONEY_REQUEST_DESCRIPTION_FORM]: OnyxTypes.Form;
    [ONYXKEYS.FORMS.MONEY_REQUEST_MERCHANT_FORM]: OnyxTypes.Form;
    [ONYXKEYS.FORMS.MONEY_REQUEST_AMOUNT_FORM]: OnyxTypes.Form;
    [ONYXKEYS.FORMS.MONEY_REQUEST_DATE_FORM]: OnyxTypes.Form;
    [ONYXKEYS.FORMS.NEW_CONTACT_METHOD_FORM]: OnyxTypes.Form;
    [ONYXKEYS.FORMS.WAYPOINT_FORM]: OnyxTypes.Form;
    [ONYXKEYS.FORMS.WAYPOINT_FORM_DRAFT]: OnyxTypes.Form;
    [ONYXKEYS.FORMS.SETTINGS_STATUS_SET_FORM]: OnyxTypes.Form;
    [ONYXKEYS.FORMS.SETTINGS_STATUS_CLEAR_AFTER_FORM]: OnyxTypes.Form;
    [ONYXKEYS.FORMS.SETTINGS_STATUS_SET_CLEAR_AFTER_FORM]: OnyxTypes.Form;
    [ONYXKEYS.FORMS.PRIVATE_NOTES_FORM]: OnyxTypes.Form;
    [ONYXKEYS.FORMS.I_KNOW_A_TEACHER_FORM]: OnyxTypes.Form;
    [ONYXKEYS.FORMS.INTRO_SCHOOL_PRINCIPAL_FORM]: OnyxTypes.Form;
```

@blazejkustra WDYT about this suggestions?

Reply:
> We have an issue here: setDraftValues() can be used for all form keys, but the draft version of these keys are not declared in ONYXKEYS.ts.

I believe we need to declare these keys in `ONYXKEYS.ts`, these are valid keys that we use and I don't see a strong reason why these should be omitted. 

> but it will slightly increase the number of keys there and could make typings more complex and slow.

I'm fine with more keys, all onyx keys that we use should be defined in `ONYXKEYS`. In terms of performance I don't think it will make a big difference.

> Another option is to move REIMBURSEMENT_ACCOUNT to inside ONYXKEYS.FORM if that is really a form.

Before typescript we had two keys `REIMBURSEMENT_ACCOUNT` and `REIMBURSEMENT_ACCOUNT_FORM` in ONYXKEYS.js. The thing is that both of these keys had the same value, so we decided to remove one of them. So we can't really move `REIMBURSEMENT_ACCOUNT` to `FORMS`.

Reply:
Ok, @kubabutkiewicz let's move with the draft keys!

> Before typescript we had two keys REIMBURSEMENT_ACCOUNT and REIMBURSEMENT_ACCOUNT_FORM in ONYXKEYS.js. The thing is that both of these keys had the same value, so we decided to remove one of them. So we can't really move REIMBURSEMENT_ACCOUNT to FORMS.

Sorry I didn't understand this @blazejkustra , why can't we move to FORMS? The value of the key won't be changed, and the FORMS object is just for better organization, isn't like COLLECTIONS which does make difference if the key is inside there or not.

Reply:
@fabioh8010 [Link](https://github.com/blazejkustra/App/pull/1#discussion_r1253244636) for more context. `REIMBURSEMENT_ACCOUNT` is used both as a form and to store reimbursement account, so technically we could move it to `ONYXKEYS.FORMS` but it's not ideal as the key is not only used as a form. 

Just noticed that there is a `REIMBURSEMENT_ACCOUNT_DRAFT` onyx key, similar to how all forms should have these drafts as well.

Reply:
@blazejkustra Thanks for the context! Just a note here, if we leave `REIMBURSEMENT_ACCOUNT` outside `FORMS` we will have to change `OnyxFormKey` to be a union of all form keys + this one. I still think would be more concise to have `REIMBURSEMENT_ACCOUNT` inside `FORMS`, but by doing that we will have refactor 20+ files and I would prefer to avoid that.

cc @kubabutkiewicz